ui/summary.go

```diff
@@ -31,99 +30,119 @@ import (
-	{"p(90)", func(s *stats.TrendSink) float64 { return s.P(0.90) }},
-	{"p(95)", func(s *stats.TrendSink) float64 { return s.P(0.95) }},
+// ErrInvalidStat represents an invalid trend column stat
+type ErrInvalidStat struct {
+	name string
+	err  error
 }
 
-type TrendColumn struct {
-	Key string
-	Get func(s *stats.TrendSink) float64
+func (e ErrInvalidStat) Error() string {
+	return errors.Wrapf(e.err, "stat '%s'", e.name).Error()
 }
 
-// VerifyTrendColumnStat checks if stat is a valid trend column
-func VerifyTrendColumnStat(stat string) error {
-	if stat == "" {
-		return ErrStatEmptyString
+// Summary handles test summary output
+type Summary struct {
+	trendColumns        []string
+	trendValueResolvers map[string]func(s *stats.TrendSink) interface{}
+}
+
+// NewSummary returns a new Summary struct. It returns an error if the
+// passed stats fields cannot be used as a trend column.
+func NewSummary(sts []string) (*Summary, error) {
+	tc := []string{"avg", "min", "med", "max", "p(90)", "p(95)"}
+	if len(sts) > 0 {
+		tc = sts
 	}
 
-	for _, col := range TrendColumns {
-		if col.Key == stat {
-			return nil
-		}
+	resolvers := map[string]func(s *stats.TrendSink) interface{}{
+		"avg":   func(s *stats.TrendSink) interface{} { return s.Avg },
+		"min":   func(s *stats.TrendSink) interface{} { return s.Min },
+		"med":   func(s *stats.TrendSink) interface{} { return s.Med },
+		"max":   func(s *stats.TrendSink) interface{} { return s.Max },
+		"p(90)": func(s *stats.TrendSink) interface{} { return s.P(0.90) },
+		"p(95)": func(s *stats.TrendSink) interface{} { return s.P(0.95) },
+		"count": func(s *stats.TrendSink) interface{} { return s.Count },
 	}
 
-	_, err := generatePercentileTrendColumn(stat)
-	return err
+	s := Summary{trendColumns: tc, trendValueResolvers: resolvers}
+	if err := s.Validate(); err != nil {
+		return nil, err
+	}
+
+	customResolvers := s.generateCustomTrendValueResolvers(sts)
+	for name, res := range customResolvers {
+		s.trendValueResolvers[name] = res
+	}
+
+	return &s, nil
 }
 
-// UpdateTrendColumns updates the default trend columns with user defined ones
-func UpdateTrendColumns(stats []string) {
-	newTrendColumns := make([]TrendColumn, 0, len(stats))
+func (s *Summary) generateCustomTrendValueResolvers(sts []string) map[string]func(s *stats.TrendSink) interface{} {
+	resolvers := make(map[string]func(s *stats.TrendSink) interface{})
+	for _, stat := range sts {
+		if _, exists := s.trendValueResolvers[stat]; !exists {
+			percentile, err := validatePercentile(stat)
+			if err == nil {
+				resolvers[stat] = func(s *stats.TrendSink) interface{} { return s.P(percentile / 100) }
+			}
+		}
+	}
 
-	for _, stat := range stats {
-		percentileTrendColumn, err := generatePercentileTrendColumn(stat)
+	return resolvers
+}
 
-		if err == nil {
-			newTrendColumns = append(newTrendColumns, TrendColumn{stat, percentileTrendColumn})
+// Validate checks if defined trend columns are valid
+func (s *Summary) Validate() error {
```

Comment:
This can technically be private, since it's called implicitly in `NewSummary()` and not used anywhere else, but validation functions should probably be public.

Reply:
I'm wondering if this shouldn't actually be a separate function, not a method of `Summary`? That way you can use it in the config handling without having to go through the `NewSummary` invocation and result discarding...

Reply:
Thanks @na--, I like that. Will change.

Reply:
Resolved by b4a776dc, but it's a bit sloppy since `getDefaultResolvers()` is now a thing. But it's certainly nicer than before. Thanks for the suggestion.

Though this file and `ui` in general need a bigger cleanup out of scope for this PR.